src/App.tsx

```diff
@@ -1,16 +1,143 @@
+  useEffect,
+  useState,
+} from 'react';
+import { Todo } from './types/Todo';
+import { User } from './types/User';
+import { Error } from './types/Error';
 import { AuthContext } from './components/Auth/AuthContext';
+import {
+  getTodos,
+  addTodo,
+  deleteTodo,
+  changeTodo,
+} from './api/todos';
+
+import { TodoList } from './components/TodoList/TodoList';
+import { NewTodo } from './components/NewTodo/NewTodo';
+import { TodoError } from './components/TodoError/TodoError';
 
 export const App: React.FC = () => {
   // eslint-disable-next-line @typescript-eslint/no-unused-vars
   const user = useContext(AuthContext);
-  const newTodoField = useRef<HTMLInputElement>(null);
+  const [userTodos, setUserTodos] = useState<Todo[]>([]);
+  const [hasTodos, setHasTodos] = useState(false);
+  const [hasError, setHasError] = useState<Error>({ status: false });
+  const [isAdding, setIsAdding] = useState(false);
+  const [tempTodo, setTempTodo] = useState<Todo | null>(null);
+  const [changingIds, setChangingIds] = useState<number[]>([0]);
+
+  const handleSetError = useCallback((error: Error) => {
+    setHasError(error);
+    setTimeout(() => {
+      setHasError({ status: false });
+    }, 3000);
+  }, []);
+
+  const handleLoadTodos = useCallback(async () => {
+    try {
+      setHasError({ status: false });
+
+      const todosFromServer = await getTodos((user as User).id);
+
+      setUserTodos(todosFromServer);
+    } catch {
+      handleSetError({ status: true });
+    }
+  }, []);
 
   useEffect(() => {
-    // focus the element with `ref={newTodoField}`
-    if (newTodoField.current) {
-      newTodoField.current.focus();
+    handleLoadTodos();
+  }, []);
+
+  useEffect(() => {
+    if (userTodos.length !== 0) {
+      setHasTodos(true);
+    } else {
+      setHasTodos(false);
+    }
+  }, [userTodos]);
+
+  const handleCloseError = useCallback(() => {
+    handleSetError({ status: false });
+  }, []);
+
+  const handleAddTodo = useCallback(async (title: string) => {
+    handleSetError({ status: false });
+
+    try {
+      const preparedData = {
+        title,
+        userId: (user as User).id,
+        completed: false,
+      };
+
+      setTempTodo({ ...preparedData, id: 0 });
+      setIsAdding(true);
+
+      await addTodo(preparedData);
+      await handleLoadTodos();
+
+      setIsAdding(false);
+    } catch {
+      handleSetError({ status: true, message: 'Unable to add a todo' });
+      setIsAdding(false);
+    }
+  }, []);
+
+  const handleDeleteTodo = useCallback(async (id: number) => {
+    handleSetError({ status: false });
+
+    try {
+      setChangingIds(currValue => [...currValue, id]);
+
+      await deleteTodo(id);
+      await handleLoadTodos();
+
+      setChangingIds(currValue => currValue.filter(currId => currId !== id));
+    } catch {
+      handleSetError({ status: true, message: 'Unable to delete a todo' });
+      setChangingIds(currValue => currValue.filter(currId => currId !== id));
```

Comment:
maybe for avoiding duplicate this code you should to use `finally` block
```suggestion 
    } catch {
      handleSetError({ status: true, message: 'Unable to delete a todo' });
    } finally {
      setChangingIds(currValue => currValue.filter(currId => currId !== id));
```

Reply:
i think it's not possible

Reply:
but why?
and even without finally method you still can move logic out of try/catch block